### PR TITLE
[Execution/Commit] Tiny optimizations.

### PR DESCRIPTION
**Pull request by grao1991**
### Description
10%~20% improvement when db is not huge, on a 128vCPU machine.

### CODE REVIEW

File: execution/executor/src/components/apply_chunk_output.rs [hunk 1]
```diff
@@ -147,23 +162,44 @@ impl ApplyChunkOutput {
...
+        let mut to_commit = Vec::with_capacity(num_txns);
+        let mut txn_info_hashes = Vec::with_capacity(num_txns);
+        let hashes_vec = {
+            let _timer = APTOS_EXECUTOR_OTHER_TIMERS_SECONDS
+                .with_label_values(&["calculate_ledger_diff_hashes"])
+                .start_timer();
+            to_keep
+                .par_iter()
+                .with_min_len(16)
+                .map(|t| {
```
Comment (msmouse) [inline execution/executor/src/components/apply_chunk_output.rs:178]: nit: let's give `t.1` a name

File: execution/executor/src/components/apply_chunk_output.rs [hunk 2]
```diff
@@ -147,23 +162,44 @@ impl ApplyChunkOutput {
...
+                            .iter()
+                            .map(CryptoHash::hash)
+                            .collect::<Vec<_>>(),
+                        CryptoHash::hash(t.1.write_set()),
+                    )
+                })
+                .collect::<Vec<_>>()
+        };
+
+        for ((txn, txn_output), state_checkpoint_hash, state_updates, hashes) in itertools::izip!(
```
Comment (msmouse) [inline execution/executor/src/components/apply_chunk_output.rs:190]: nit: ditto, name the hashes

File: storage/aptosdb/src/lib.rs [hunk 3]
```diff
@@ -1584,6 +1595,7 @@ impl DbWriter for AptosDB {
                     let _timer = OTHER_TIMERS_SECONDS
                         .with_label_values(&["save_transactions_commit"])
                         .start_timer();
+                    info!("{:?}", Instant::now());
```
Comment (msmouse) [inline storage/aptosdb/src/lib.rs:1598]: remove or proper message

File: storage/schemadb/src/lib.rs [hunk 4]
```diff
@@ -221,6 +221,7 @@ impl DB {
 
     /// Writes a group of records wrapped in a [`SchemaBatch`].
     pub fn write_schemas(&self, batch: SchemaBatch) -> Result<()> {
+        info!("{:?}, {}", std::time::Instant::now(), self.name);
```
Comment (msmouse) [inline storage/schemadb/src/lib.rs:224]: remove or proper message

**Comment by msmouse** [comment]
1. These micro- parallelism bothers me.. but I can't say it's simply wrong as long as you can prove it improves performance.. (however does it really help in "real life" where the state is reasonably large and hitting disk?
2. The concurrency in save_transactions(), same deal, I doubt it matters when the state is reasonably large and hitting disk. But well, this doesn't bother me as much because it's only concurrency of two.

**Review by msmouse — APPROVED** [review]

**Review by msmouse — CHANGES_REQUESTED** [review]
accidental approval

File: execution/executor/src/components/apply_chunk_output.rs (hunk 1 above)
Comment (grao1991) [reply execution/executor/src/components/apply_chunk_output.rs:178]: Done

File: execution/executor/src/components/apply_chunk_output.rs (hunk 2 above)
Comment (grao1991) [reply execution/executor/src/components/apply_chunk_output.rs:190]: Done

File: storage/aptosdb/src/lib.rs (hunk 3 above)
Comment (grao1991) [reply storage/aptosdb/src/lib.rs:1598]: Removed.

File: storage/schemadb/src/lib.rs (hunk 4 above)
Comment (grao1991) [reply storage/schemadb/src/lib.rs:224]: Removed.

**Comment by grao1991** [comment]
> 1. These micro- parallelism bothers me.. but I can't say it's simply wrong as long as you can prove it improves performance.. (however does it really help in "real life" where the state is reasonably large and hitting disk?
> 2. The concurrency in save_transactions(), same deal, I doubt it matters when the state is reasonably large and hitting disk. But well, this doesn't bother me as much because it's only concurrency of two.

I think as long as the benefit is larger than the overhead of the context switch, and there is unused CPU at any given stage, it's worth to have extra threads.

File: storage/aptosdb/src/state_store/mod.rs [hunk 5]
```diff
@@ -768,9 +783,10 @@ impl StateValueWriter<StateKey, StateValue> for StateStore {
     }
 }
 
-fn add_kv_batch(batch: &mut SchemaBatch, kv_batch: &StateValueBatch) -> Result<()> {
-    for (k, v) in kv_batch {
-        batch.put::<StateValueSchema>(k, v)?;
-    }
+fn add_kv_batch(batch: &SchemaBatch, kv_batch: &StateValueBatch) -> Result<()> {
+    kv_batch
+        .par_iter()
```
Comment (zekun000) [inline storage/aptosdb/src/state_store/mod.rs:788]: without min_len?

**Review by zekun000 — APPROVED** [review]

**Review by msmouse — APPROVED** [review]
Why doesn't the  forge run on the PR show a 10%-20% improvement you described?

**Comment by grao1991** [comment]
> Why doesn't the forge run on the PR show a 10%-20% improvement you described?

Forge is tricky. I did my experiment on a 128vCPU machine with executor-benchmark (almost non-conflicting txns)

File: storage/aptosdb/src/state_store/mod.rs (hunk 5 above)
Comment (grao1991) [reply storage/aptosdb/src/state_store/mod.rs:788]: Doesn't work on map iterator..

BTW I moved the commit code path away from this function, to save a clone. This function now is only used by restore path.

**Pull request merged**